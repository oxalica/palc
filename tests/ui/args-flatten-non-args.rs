//! TODO: This triggers are many duplicated errors. Could be improved, but blocked on rustc.
//! WAIT: <https://github.com/rust-lang/rust/issues/144745>

#[derive(palc::Args)]
struct Cli1 {
    #[command(flatten)]
    deep: Deep,
}

struct Deep {}

#[derive(palc::Args)]
struct Cli2 {
    #[command(flatten)]
    deep: Subcmd,
}

#[derive(palc::Subcommand)]
enum Subcmd {}

fn main() {}
