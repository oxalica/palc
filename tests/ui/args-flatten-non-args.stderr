error[E0277]: cannot flatten `Deep` which is not a `palc::Args`
 --> tests/ui/args-flatten-non-args.rs:7:11
  |
7 |     deep: Deep,
  |           ^^^^ this type is expected to have `derive(palc::Args)` but it is not
  |
  = help: the trait `palc::__private::Args` is not implemented for `Deep`
  = help: the following other types implement trait `palc::__private::Args`:
            Cli1
            Cli2

error[E0277]: cannot flatten `Deep` which is not a `palc::Args`
 --> tests/ui/args-flatten-non-args.rs:5:8
  |
5 | struct Cli1 {
  |        ^^^^ this type is expected to have `derive(palc::Args)` but it is not
  |
  = help: within `Cli1State`, the trait `palc::__private::Args` is not implemented for `Deep`
  = help: the following other types implement trait `palc::__private::Args`:
            Cli1
            Cli2
note: required because it appears within the type `Cli1State`
 --> tests/ui/args-flatten-non-args.rs:5:8
  |
5 | struct Cli1 {
  |        ^^^^
note: required by a bound in `palc::__private::Args::__State`
 --> src/runtime.rs
  |
  |     type __State: ParserState<Output = Self>;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Args::__State`

error[E0277]: cannot flatten `Subcmd` which is not a `palc::Args`
  --> tests/ui/args-flatten-non-args.rs:15:11
   |
15 |     deep: Subcmd,
   |           ^^^^^^ this type is expected to have `derive(palc::Args)` but it is not
   |
   = help: the trait `palc::__private::Args` is not implemented for `Subcmd`
   = help: the following other types implement trait `palc::__private::Args`:
             Cli1
             Cli2

error[E0277]: cannot flatten `Subcmd` which is not a `palc::Args`
  --> tests/ui/args-flatten-non-args.rs:13:8
   |
13 | struct Cli2 {
   |        ^^^^ this type is expected to have `derive(palc::Args)` but it is not
   |
   = help: within `Cli2State`, the trait `palc::__private::Args` is not implemented for `Subcmd`
   = help: the following other types implement trait `palc::__private::Args`:
             Cli1
             Cli2
note: required because it appears within the type `Cli2State`
  --> tests/ui/args-flatten-non-args.rs:13:8
   |
13 | struct Cli2 {
   |        ^^^^
note: required by a bound in `palc::__private::Args::__State`
  --> src/runtime.rs
   |
   |     type __State: ParserState<Output = Self>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Args::__State`

error[E0277]: cannot flatten `Deep` which is not a `palc::Args`
 --> tests/ui/args-flatten-non-args.rs:4:10
  |
4 | #[derive(palc::Args)]
  |          ^^^^^^^^^^ this type is expected to have `derive(palc::Args)` but it is not
  |
  = help: within `Cli1State`, the trait `palc::__private::Args` is not implemented for `Deep`
  = help: the following other types implement trait `palc::__private::Args`:
            Cli1
            Cli2
note: required because it appears within the type `Cli1State`
 --> tests/ui/args-flatten-non-args.rs:5:8
  |
5 | struct Cli1 {
  |        ^^^^
  = note: the return type of a function must have a statically known size
  = note: this error originates in the derive macro `palc::Args` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: cannot flatten `Subcmd` which is not a `palc::Args`
  --> tests/ui/args-flatten-non-args.rs:12:10
   |
12 | #[derive(palc::Args)]
   |          ^^^^^^^^^^ this type is expected to have `derive(palc::Args)` but it is not
   |
   = help: within `Cli2State`, the trait `palc::__private::Args` is not implemented for `Subcmd`
   = help: the following other types implement trait `palc::__private::Args`:
             Cli1
             Cli2
note: required because it appears within the type `Cli2State`
  --> tests/ui/args-flatten-non-args.rs:13:8
   |
13 | struct Cli2 {
   |        ^^^^
   = note: the return type of a function must have a statically known size
   = note: this error originates in the derive macro `palc::Args` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: cannot flatten `Deep` which is not a `palc::Args`
 --> tests/ui/args-flatten-non-args.rs:4:10
  |
4 | #[derive(palc::Args)]
  |          ^^^^^^^^^^ this type is expected to have `derive(palc::Args)` but it is not
  |
  = help: the trait `palc::__private::Args` is not implemented for `Deep`
  = help: the following other types implement trait `palc::__private::Args`:
            Cli1
            Cli2
  = note: this error originates in the derive macro `palc::Args` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: cannot flatten `Subcmd` which is not a `palc::Args`
  --> tests/ui/args-flatten-non-args.rs:12:10
   |
12 | #[derive(palc::Args)]
   |          ^^^^^^^^^^ this type is expected to have `derive(palc::Args)` but it is not
   |
   = help: the trait `palc::__private::Args` is not implemented for `Subcmd`
   = help: the following other types implement trait `palc::__private::Args`:
             Cli1
             Cli2
   = note: this error originates in the derive macro `palc::Args` (in Nightly builds, run with -Z macro-backtrace for more info)
